hack/test-go.sh

```diff
@@ -1,125 +1,231 @@
 #!/bin/bash
-
-# See HACKING.md for usage
+#
+# This script runs Go language unit tests for the Origin repository. Arguments to this script
+# are parsed as a list of packages to test. If none are given, then "all" packages are tested. 
+# Coverage reports and jUnit XML reports can be generated by this script as well, but both cannot
+# be generated at once.
+#
+# This script consumes the following parameters as environment variables:
+#  - PRINT_PACKAGES:      prints all packages that would be tested and exits, dry run
+#  - TEST_KUBE:           toggles testing of non-essential Kubernetes unit tests
+#  - TIMEOUT:             the timeout for any one unit test (default '60s')
+#  - DETECT_RACES:        toggles the 'go test' race detector (defaults '-race')
+#  - COVERAGE_OUTPUT_DIR: locates the directory in which coverage output files will be placed
+#  - COVERAGE_SPEC:       a set of flags for 'go test' that specify the coverage behavior (default '-cover -covermode=atomic')
+#  - GOTEST_FLAGS:        any other flags to be sent to 'go test'
+#  - JUNIT_REPORT:        toggles the creation of jUnit XML from the test output and changes this script's output behavior
+#                         to use the 'junitreport' tool for summarizing the tests.
 
 set -o errexit
 set -o nounset
 set -o pipefail
 
-STARTTIME=$(date +%s)
-OS_ROOT=$(dirname "${BASH_SOURCE}")/..
-source "${OS_ROOT}/hack/common.sh"
+function exit_trap() {
+    local return_code=$?
 
-# Go to the top of the tree.
-cd "${OS_ROOT}"
+    end_time=$(date +%s)
+    
+    if [[ "${return_code}" -eq "0" ]]; then
+        verb="succeeded"
+    else
+        verb="failed"
+    fi
 
-PRINT_PACKAGES=${PRINT_PACKAGES:-""}
-TEST_KUBE=${TEST_KUBE:-""}
-KUBE_GODEP_PATH="./Godeps/_workspace/src/k8s.io/kubernetes/pkg"
+    echo "$0 ${verb} after $((${end_time} - ${start_time})) seconds"
+    exit "${return_code}"
+}
+
+trap exit_trap EXIT
 
+start_time=$(date +%s)
```

Comment:
I think I'll stick with `date +%s` for now to keep this parallel with other scripts - we can fix this in the trap PR.